examples/NAT.cs

```diff
@@ -36,223 +36,348 @@
-    from.tcp_port = p_tcp.SourcePort;
+#if DEBUG
+        Console.WriteLine("RX {0}:{1} -> {2}:{3} on {4} [{5}]",
+          p_ip.SourceAddress, p_tcp.SourcePort, p_ip.DestinationAddress, p_tcp.DestinationPort, in_port,
+          p_tcp.Syn?"S":"");
+#endif
 
-    int out_port = Port_Drop;
-    if (in_port == Port_Outside)
-    {
-      from.assigned_tcp_port = p_tcp.DestinationPort;
-      from.network_port = null;
-      out_port = outside_to_inside (p_ip, p_tcp, from);
-    } else {
-      from.assigned_tcp_port = null;
-      from.network_port = in_port;
-      out_port = inside_to_outside (p_ip, p_tcp, from);
-    }
+        int out_port = Port_Drop;
+        if (in_port == Port_Outside)
+          out_port = outside_to_inside (p_eth, p_ip, p_tcp);
+        else
+          out_port = inside_to_outside (p_eth, p_ip, p_tcp, in_port);
 
 #if DEBUG
-    print_mapping (port_mapping, " (------ Outside ------) ", " (------ Inside ------) ");
-    Console.WriteLine();
-    print_mapping (port_reverse_mapping, " (------ Inside ------) ", " (------ Outside ------) ");
+        Console.WriteLine(" (------ Outside ------)  \t<->\t (------ Inside ------) ");
+        foreach (var entry in NAT_MapToInside)
+        {
+          Console.WriteLine("{0} \t<->\t {1}", entry.Key, entry.Value);
+        }
+        Console.WriteLine();
+        Console.WriteLine(" (------ Inside ------)  \t<->\t (------ Outside ------) ");
+        foreach (var entry in NAT_MapToOutside)
+        {
+          Console.WriteLine("{0} \t<->\t {1}", entry.Key, entry.Value);
+        }
 #endif
 
-    return out_port;
+        return out_port;
+      }
+    }
+
+    // Drop packets that aren't handled
+    return Port_Drop;
   }
 
-  private int outside_to_inside (IpPacket p_ip, TcpPacket p_tcp, NAT_Entry from)
+  /// <summary>
+  /// Rewrite packets coming from the Outside and forward on the relevant Inside network port.
+  /// </summary>
+  private int outside_to_inside (EthernetPacket p_eth, IpPacket p_ip, TcpPacket p_tcp)
   {
+    // p_ip.DestinationAddress should be my_address
+    if (!p_ip.DestinationAddress.Equals(my_address))
+      return Port_Drop;
+
     // Retrieve the mapping. If a mapping doesn't exist, then it means that we're not
     // aware of a session to which the packet belongs: so drop the packet.
-    NAT_Entry to;
-    if (port_mapping.TryGetValue(from, out to))
+    InternalNode destination;
+    var key = new MapToInside_Key(p_ip.SourceAddress, p_tcp.SourcePort, p_tcp.DestinationPort);
+    if (NAT_MapToInside.TryGetValue(key, out destination))
     {
       // Rewrite destination IP address and TCP port, and map to the appropriate Inside port.
-      p_ip.DestinationAddress = to.ip_address;
-      p_tcp.DestinationPort = to.tcp_port;
-      // Update checksums.
-      p_tcp.UpdateTCPChecksum();
+      p_ip.DestinationAddress = destination.Address;
+      p_tcp.DestinationPort = destination.Port;
+      // Update checksums. NOTE IP updated before TCP
```

Comment:
Makes sense, thanks. I agree that the checksum computation should be 
commutative for IP and TCP -- i.e., you shouldn't see any difference if 
the order of checksum computations (i.e., checksum of IP first and TCP 
second) were swapped, because the TCP's pseudoheader doesn't include 
IP's checksum. Out of curiosity, did you notice anything break when the 
order was swapped? (i.e., checksum of TCP first and IP second)

On 2016-07-07 12:47, TMVector wrote:
> In examples/NAT.cs [1]:
> 
>> {
>> // Rewrite destination IP address and TCP port, and map to the
> appropriate Inside port.
>> - p_ip.DestinationAddress = to.ip_address;
>> - p_tcp.DestinationPort = to.tcp_port;
>> - // Update checksums.
>> - p_tcp.UpdateTCPChecksum();
>> + p_ip.DestinationAddress = destination.Address;
>> + p_tcp.DestinationPort = destination.Port;
>> + // Update checksums. NOTE IP updated before TCP
> 
> The IPv4 checksum depends only on the header fields, not on the
> payload, so it _can_ be done first. However the TCP checksum is
> computed over the TCP header, TCP payload, AND the _psuedoheader_:
> 
> From RFC 793 [2]:
> 
>> ...
>> The checksum also covers a 96 bit pseudo header conceptually
>> prefixed to the TCP header. This pseudo header contains the Source
>> Address, the Destination Address, the Protocol, and TCP length. This
>> gives the TCP protection against misrouted segments. This
>> information is carried in the Internet Protocol and is transferred
>> across the TCP/Network interface in the arguments or results of
>> calls by the TCP on the IP.
>> ...
> 
> So the TCP checksum relies on some IPv4 information, but I'm not sure
> why this means that the IPv4 checksum _has_ to be computed first, as
> the TCP pseudo header doesn't include the IPv4 checksum. From a quick
> glance over the PacketDotNet source code I can't see any issues
> either.
> 
> --
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub [3], or mute the
> thread [4].
> 
> 	*
> 
> Links:
> ------
> [1] https://github.com/niksu/pax/pull/3#discussion_r69892716
> [2] https://tools.ietf.org/html/rfc793#page-16
> [3]
> https://github.com/niksu/pax/pull/3/files/d721adebf668133ae20c55f466ccfaa2344f51a2#r69892716
> [4]
> https://github.com/notifications/unsubscribe/ABDTTgdIHuVbY2Saawiw83_Sz7BjyTY-ks5qTOdsgaJpZM4JFDoJ

-- 
http://www.cl.cam.ac.uk/~ns441/